### DISCUSSION TITLE
identifier "false" is undefined

### DISCUSSION THREAD

**Discussion by winstxnhdw** [microsoft/vscode-cpptools#11718] [Q&A] [answered]
I have Clang 16 installed and I can compile my code just fine, but intellisense says that my booleans are undefined.

```json
{
  "configurations": [
    {
      "name": "Linux",
      "includePath": ["/usr/lib/clang/16/**", "${workspaceFolder}/**"],
      "defines": [],
      "compilerPath": "/sbin/clang-16",
      "intelliSenseMode": "linux-clang-x64",
      "cStandard": "c23",
      "cppStandard": "c++23"
    }
  ],
  "version": 4
}
```

Any ideas?

**Comment by bobbrow** [comment]
It is not recommended to add the system include path to your `includePath` setting. The extension should be able to determine that by querying the compiler listed in the `compilerPath` setting.

In the case where you're seeing this error could you run the "C/C++: Log Diagnostics" command when the affected file is visible in the editor and share the output with us?

**Reply by winstxnhdw** [reply]
Yeah, I tried both configurations. I just added that to see if it would resolve the problem (which it didn't).

```
-------- Diagnostics - 21/11/2023, 18:31:26
Version: 1.18.5
Current Configuration:
{
    "name": "Linux",
    "includePath": [
        "/home/winston/projects/libcs1010/**"
    ],
    "defines": [],
    "compilerPath": "/sbin/clang-16",
    "intelliSenseMode": "linux-clang-x64",
    "cStandard": "c23",
    "cppStandard": "c++23",
    "compilerPathIsExplicit": true,
    "cStandardIsExplicit": true,
    "cppStandardIsExplicit": true,
    "intelliSenseModeIsExplicit": true,
    "compilerPathInCppPropertiesJson": "/sbin/clang-16",
    "mergeConfigurations": false,
    "browse": {
        "path": [
            "/home/winston/projects/libcs1010/**",
            "${workspaceFolder}"
        ],
        "limitSymbolsToIncludedHeaders": true
    }
}
cpptools version (native): 1.18.3.0
Translation Unit Mappings:
[ /home/winston/projects/libcs1010/main.c ]:
    /home/winston/projects/libcs1010/main.c
    /home/winston/projects/libcs1010/include/cs1010.h *
Translation Unit Configurations:
[ /home/winston/projects/libcs1010/main.c ]:
    Process ID: 47182
    Memory Usage: 19 MB
    Compiler Path: /sbin/clang-16
    Includes:
        /usr/lib/clang/16/include
        /usr/local/include
        /usr/include
        /home/winston/projects/libcs1010/include
    Standard Version: c23
    IntelliSense Mode: linux-clang-x64
    Other Flags:
        --clang
        --clang_version=160006
Total Memory Usage: 19 MB

------- Workspace parsing diagnostics -------
Number of files discovered (not excluded): 22581
```

**Reply by bobbrow** [reply, marked as answer]
Ah, ok. You're using C.  Support for `true` and `false` in the C23 spec has not yet been implemented by the language server.  Lists of supported features per standard are linked from here: https://www.edg.com/c/features.  This specific one is on row 44 of the spreadsheet for C23 feature tracking.

**Reply by winstxnhdw** [reply]
😂😂😂 and I am guessing there’s no alternative either.

**Reply by bobbrow** [reply]
Well, as a workaround, for the time being you could add `"true=1"` and `"false=0"` to the `"defines"` array in your `c_cpp_properties.json`.

**Reply by HosseinSagha** [reply]
workaround won't work.

**Reply by bobbrow** [reply]
@HosseinSagha can you share how you added the workaround into your c_cpp_properties.json?  It seems to be working for 6 other people who upvoted the workaround.

**Reply by HosseinSagha** [reply]
Arm gnu toolchain 14.3.1


<img width="172" height="101" alt="image" src="https://github.com/user-attachments/assets/f1cdf1e2-4069-4c91-8436-447a132db3d3" />


<img width="287" height="512" alt="image" src="https://github.com/user-attachments/assets/92262ecc-59ac-4bc8-a34d-f72d1ed1cc31" />

**Reply by bobbrow** [reply]
Thank you. Can you please try adding `"mergeConfigurations": true` to your configuration? When you use CMake Tools to configure your project, we ignore the other settings by default because they might conflict.

**Reply by HosseinSagha** [reply]
now, it worked. Thank you
<img width="269" height="419" alt="image" src="https://github.com/user-attachments/assets/591c49d3-cf9b-4013-9fef-b25ac238ff7b" />

**Comment by starball5** [comment]
Related on Stack Overflow: [VS Code's C/C++ extension says the C23 true and false keywords are unidentified](https://stackoverflow.com/q/77830499/11107541).
